resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java

```diff
@@ -201,50 +202,56 @@ private State calculateNextState(final long timeoutInNanos, final State activeSt
             nextPermissions = (int) min(nextPermissions + accumulatedPermissions, permissionsPerCycle);
         }
         long nextNanosToWait = nanosToWaitForPermission(
-                cyclePeriodInNanos, permissionsPerCycle, nextPermissions, currentNanos, currentCycle
+                weight, cyclePeriodInNanos, permissionsPerCycle, nextPermissions, currentNanos, currentCycle
         );
-        State nextState = reservePermissions(activeState.config, timeoutInNanos, nextCycle, nextPermissions, nextNanosToWait);
+        State nextState = reservePermissions(activeState.config, weight, timeoutInNanos, nextCycle, nextPermissions, nextNanosToWait);
         return nextState;
     }
 
     /**
-     * Calculates time to wait for next permission as
-     * [time to the next cycle] + [duration of full cycles until reserved permissions expire]
+     * Calculates time to wait for the required weight of permissions to get accumulated
      *
+     * @param weight               weight of required permissions
      * @param cyclePeriodInNanos   current configuration values
      * @param permissionsPerCycle  current configuration values
      * @param availablePermissions currently available permissions, can be negative if some permissions have been reserved
      * @param currentNanos         current time in nanoseconds
      * @param currentCycle         current {@link AtomicRateLimiter} cycle    @return nanoseconds to wait for the next permission
      */
-    private long nanosToWaitForPermission(final long cyclePeriodInNanos, final int permissionsPerCycle,
+    private long nanosToWaitForPermission(final int weight, final long cyclePeriodInNanos, final int permissionsPerCycle,
                                           final int availablePermissions, final long currentNanos, final long currentCycle) {
-        if (availablePermissions > 0) {
+        if (availablePermissions >= weight) {
             return 0L;
         }
         long nextCycleTimeInNanos = (currentCycle + 1) * cyclePeriodInNanos;
         long nanosToNextCycle = nextCycleTimeInNanos - currentNanos;
-        int fullCyclesToWait = (-availablePermissions) / permissionsPerCycle;
-        return (fullCyclesToWait * cyclePeriodInNanos) + nanosToNextCycle;
+        int permissionsToTakeAfterCurrencyCycle = weight - availablePermissions;
+        int fullCyclesToWaitCount = Math.floorDiv(permissionsToTakeAfterCurrencyCycle, permissionsPerCycle);
+        if (permissionsToTakeAfterCurrencyCycle % permissionsPerCycle == 0) {
+            // will take all permissions from the last cycle but can reserve them without waiting till its end
+            fullCyclesToWaitCount -= 1;
+        }
+        return nanosToNextCycle + (fullCyclesToWaitCount * cyclePeriodInNanos);
```

Comment:
I'd suggest such simplification
```java
int permissionsAtTheStartOfNextCycle = availablePermissions + permissionsPerCycle;
int fullCyclesToWait = -(permissionsAtTheStartOfNextCycle-weight) / permissionsPerCycle;
return (fullCyclesToWait * cyclePeriodInNanos) + nanosToNextCycle;
```
I'm not totally sure about all corner cases, but all tests appear to work OK.
If I missed something, please correct me.

Reply:
ok, done